File: 787. Cheapest Flights Within K Stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, K: int) -> int:
        '''
        key point Dijkstra's algorithm,
        we need to put (price, steps left, pos) in the heap
        and keep tracking steps left, if we have tried all possible path and we still get no answer,
        we cannot go through
        and the first answer we meet should be the optimal solution
        '''
        graph = collections.defaultdict(dict)
        for a, b, price in flights:
            graph[a][b] = price
        
        pq = [(0, K+1, src)]
        while pq:
            price, k, pos = heapq.heappop(pq)
            if pos == dst:
                return price
            if k > 0:
                newpos = graph[pos]
                for p in newpos:
                    heapq.heappush(pq, (price+newpos[p], k-1, p))
        return -1

File: 787. Cheapest Flights Within K Stops.java
class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int K) {
        int[][] graph = new int[n][n];
        for (int[] fl : flights) {
            graph[fl[0]][fl[1]] = fl[2];
        }
        PriorityQueue<Tuple> pq = new PriorityQueue<>();
        pq.offer(new Tuple(src, 0, 0));
        while (!pq.isEmpty()) {
            Tuple t = pq.poll();
            if (t.stop-1 > K) {continue;}
            if (t.city == dst) {return t.price;}
            for (int i=0; i<n; i++) {
                if (graph[t.city][i] > 0) {
                    pq.offer(new Tuple(i, t.stop+1, t.price+graph[t.city][i]));
                }
            }
        }
        return -1;
    }
    
    class Tuple implements Comparable<Tuple> {
        int city;
        int stop;
        int price;
        public Tuple(int city, int stop, int price) {
            this.city = city;
            this.stop = stop;
            this.price = price;
        }
        @Override
        public int compareTo(Tuple that) {
            return this.price - that.price;
        }
    }
}